packages/next-auth/src/providers/shopify.js

```diff
@@ -0,0 +1,79 @@
+export default function Shopify(options) {
+  const {
+    shop = "",
+    scope = "",
+    clientId = "",
+    clientSecret = "",
+    apiVersion = "",
+  } = options
+
+  return {
+    id: "shopify",
+    name: "Shopify",
+    type: "oauth",
+    authorization: {
+      url: `https://${shop}.myshopify.com/admin/oauth/authorize`,
+      params: {
+        scope,
+      },
+    },
+    token: {
+      request: async (context) => {
+        if (!context.params.code)
+          throw new Error("No code returned by shopify server")
+        const payload = JSON.stringify({
+          client_id: clientId,
+          client_secret: clientSecret,
+          code: context.params.code,
+        })
+        const res = await fetch(
+          `https://${shop}.myshopify.com/admin/oauth/access_token`,
+          {
+            method: "POST",
+            headers: {
+              Accept: "application/json",
+              "Content-Type": "application/json",
+            },
+            body: payload,
+          }
+        )
+        if (res.ok) {
+          const result = await res.json()
+          return { tokens: result }
+        }
+
+        throw new Error(
+          "Something went wrong while trying to get the access_token"
+        )
+      },
+    },
+    userinfo: {
+      request: async ({ tokens }) => {
+        const { access_token = "" } = tokens
+        if (!access_token) throw new Error("Access token is missing")
+        const res = await fetch(
+          `https://${shop}.myshopify.com/admin/api/${apiVersion}/shop.json`,
+          {
+            headers: {
+              "Content-Type": "application/json",
+              "X-Shopify-Access-Token": access_token,
+            },
+          }
+        )
+        if (res.ok) {
+          const result = await res.json()
+          return result
+        }
+
+        throw new Error("Something went wrong while trying to access your shop")
+      },
+    },
```

Comment:
```suggestion
    userinfo: `https://${shop}.myshopify.com/admin/api/${apiVersion}/shop.json`,
```

Potentially this should be enough, although it would need verification due to this I found: https://community.shopify.com/c/shopify-apis-and-sdks/x-shopify-access-token-vs-authorization-bearer/m-p/564674

Looks like Shopify is claiming OAuth 2.0 compliance but they are skipping this step...

https://datatracker.ietf.org/doc/html/rfc6749#section-7.1

Or they might allow it now, could you just verify, please?

Reply:
Shopify is ignoring this step [https://datatracker.ietf.org/doc/html/rfc6749#section-7.1](https://datatracker.ietf.org/doc/html/rfc6749#section-7.1).

Shopify is using `x-shopify-access-token` to fetch user-app related data and not an `authorization` header. So, we have to implement the userinfo -> request method from scratch.

Reply:
Thanks. I took this up with the Shopify Engineering team. Let's hope for a good answer.

Reply:
Any update on this? Keen to get this working, happy to assist wherever required.

In playing around with this, I've noticed a few things.

# Access Mode

Shopify has two different access modes in their auth flow: `online` and `offline` (defaults to `offline`).

Adding `grant_options[]` (as shown below) to the params in the authorization endpoint turns the `online` mode on.

```js
{
...
authorization: {
 url: `https://${shop}.myshopify.com/admin/oauth/authorize`,
 params: {
  "grant_options[]": "per-user",
  scope: "read_products",
 },
},
...
}
```

### `offline` response

```js
{
  access_token: 'shpat_xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx',
  scope: '...'
}
```

### `online` response

```js
{
  access_token: 'shpat_xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx',
  scope: '...',
  expires_at: 1645954815,
  associated_user_scope: '...',
  session: '...',
  account_number: 0,
  associated_user: {
    id: 12345,
    first_name: '...',
    last_name: '...',
    email: '...',
    account_owner: false,
    locale: 'en-AU',
    collaborator: false,
    email_verified: true
  }
}
```

`online` mode breaks adapters as the data returned no longer conforms with the NextAuth [Models](https://next-auth.js.org/adapters/models). (Testing done with Prisma adapter)

As far as I can tell this is the only way to get the actual logged-in user. Shopify's User API seems to be locked behind Shopify Plus [[reference](https://shopify.dev/api/admin-rest/2022-01/resources/user#top)]. But the `associated_user.id` value is **not** global across different Shopify instances... so a user has multiple IDs across one email.

# `userinfo`

If using the offline mode token, this results in a staff user being able to bypass scopes if they have access to the configured App in Shopify. The current setup in the `userinfo.request` callback using the `https://${shop}.myshopify.com/admin/api/${apiVersion}/shop.json` endpoint returns information on the Shopify shop. This means all staff on the Shopify instance authenticate to the same user.

# `store_name` aka `shop`

All endpoints in the Shopify Admin API follow this pattern:
https://{store_name}.myshopify.com/admin/api/{apiVersion}/{resource}.json

Requests made to Shopify must be made via the Shopify instance (as in `abc.myshopify.com`, where `abc` is the `store_name`). This must be stored somewhere for Public Apps. As far as I can tell this can never be changed [reference](https://help.shopify.com/en/manual/your-account/manage-account). 

In testing, I was using this temporary configuration for the profile callback to store the store_name as the users' name:

```js
profile: (profile) => {
 return {
  id: profile.shop.id,
  name: profile.shop.myshopify_domain,
  email: profile.shop.email,
  image: null,
 };
},
```

Note the `myshopify_domain`. This is the `store_name` and what I was using to make requests to the Shopify Admin API with. It might be wise to store the shop somewhere semantic.

Hope this helps!

An aside from this, the shop query param gets lost in the auth flow after the initial redirect from Shopify to your app. Following the advice in #3776, I pull the `shop` variable from the query params and add it to the cookie session so I can use it between the requests of the auth flow. Not great, but it works. This implementation also assumes that the user comes from Shopify, and doesn't navigate directly resulting in the shop variable being undefined :\